runtime/server/x86/frontend/fbank.h

```diff
@@ -0,0 +1,212 @@
+            first_index = i;
+          last_index = i;
+        }
+      }
+      assert(first_index != -1 && last_index >= first_index);
+      bins_[bin].first = first_index;
+      int size = last_index + 1 - first_index;
+      bins_[bin].second.resize(size);
+      for (int i = 0; i < size; i++) {
+        bins_[bin].second[i] = this_bin[first_index + i];
+      }
+    }
+
+    // hamming window
+    hamming_window_.resize(frame_length_);
+    double a = M_2PI / (frame_length-1);
+    for (int i = 0; i < frame_length; i++) {
+        hamming_window_[i] = 0.54 - 0.46*cos(a * i);
+    }
+  }
+
+  void SetUseLog(bool use_log) {
+    use_log_ = use_log;
+  }
+
+  void SetRemoveDcOffset(bool remove_dc_offset) {
+    remove_dc_offset_ = remove_dc_offset;
+  }
+
+  void SetDither(float dither) {
+    dither_ = dither;
+  }
+
+  int NumBins() const { return num_bins_; }
+
+  static inline float InverseMelScale(float mel_freq) {
+    return 700.0f * (expf (mel_freq / 1127.0f) - 1.0f);
+  }
+
+  static inline float MelScale(float freq) {
+    return 1127.0f * logf (1.0f + freq / 700.0f);
+  }
+
+  int UpperPowerOfTwo(int n) const {
+    return static_cast<int>(pow(2, ceil(log(n) / log(2))));
+  }
+
+  // preemphasis
+  void PreEmphasis(float coeff, std::vector<float>* data) const {
+    if (coeff == 0.0) return;
+    for (int i = data->size() - 1; i > 0; i--)
+      (*data)[i] -= coeff * (*data)[i-1];
+    (*data)[0] -= coeff * (*data)[0];
+  }
+
+  // Apply hamming window on data in place
+  void Hamming(std::vector<float>* data) const {
+    assert(data->size() >= hamming_window_.size());
+    for (size_t i = 0; i < hamming_window_.size(); i++) {
+      (*data)[i] *= hamming_window_[i];
+    }
+  }
+
+  // Compute fbank feat, return num frames
+  int Compute(const std::vector<float>& wave,
+              std::vector<std::vector<float> >* feat) {
+    int num_samples = wave.size();
+    if (num_samples < frame_length_) return 0;
+    int num_frames = 1 + ((num_samples - frame_length_) / frame_shift_);
+    feat->resize(num_frames);
+    std::vector<float> fft_real(fft_points_, 0), fft_img(fft_points_, 0);
+    std::vector<float> power(fft_points_ / 2);
+    for (int i = 0; i < num_frames; i++) {
+      std::vector<float> data(wave.data() + i * frame_shift_,
+                              wave.data() + i * frame_shift_ + frame_length_);
+      // optional add noise
+      if (dither_ != 0.0) {
+        for (size_t j = 0; j < data.size(); j++)
+          data[j] += dither_ * distribution_(generator_);
+      }
+      // optinal remove dc offset
+      if (remove_dc_offset_) {
+        float mean = 0.0;
+        for (size_t j = 0; j < data.size(); j++) mean += data[j];
+        mean /= data.size();
+        for (size_t j = 0; j < data.size(); j++) data[j] -= mean;
+      }
+
+      PreEmphasis(0.97, &data);
+      Hamming(&data);
+      // copy data to fft_real
+      memset(fft_real.data(), 0, sizeof(float) * fft_points_);
```

Comment:
```cpp
memset(fft_real.data() + frame_length_, 0, sizeof(float) * (fft_points_ - frame_length_));
```

Reply:
why should we subscribe frame_length here?

Reply:
Because it is `fft_real.data() + frame_length_`. Otherwise, you may get segfault.

Reply:
fixed